halo2_gadgets/src/utilities.rs

```diff
@@ -59,6 +78,62 @@ pub trait UtilitiesInstructions<F: FieldExt> {
     }
 }
 
+/// A type representing a range-constrained field element.
+#[derive(Clone, Copy, Debug)]
+pub struct RangeConstrained<F: Field, T: FieldValue<F>> {
```

Comment:
Could this be something like 

```rust
struct RangeConstrained<F: Field, NUM_BITS: usize>(AssignedCell<Bits<NUM_BITS>, F)
```

Where `Bits<NUM_BITS: usize>: Into<Assigned<F>>`.

Reply:
Something like in https://github.com/zcash/halo2/pull/525/commits/c4cebc675129ccb578170d4febb80a1586a70cfa.

Reply:
That would not work for the way the `MessagePiece::from_subpieces` API works, as we can't yet have an iterator that is generic over a type parameter. And we can't do type-system concatenation until Rust implements const generics expressions, so we'd instead need to use something like `generic_array`. It might work if we also rework the Sinsemilla APIs, but that will have to wait until after `halo2_gadgets 0.1.0` (so won't ever be used by the Orchard circuit).